ord_schema/interface/query.py

```diff
@@ -20,15 +47,252 @@
+            'components': [predicate.json() for predicate in self._predicates],
+        })
 
-class Predicate:
-    """Structure and code generation for ORD query predicates."""
+    def _setup(self, cursor):
+        """Prepares the database for a query.
+
+        Args:
+            cursor: psycopg.cursor instance.
+        """
+        command = sql.SQL('SET rdkit.do_chiral_sss=%s')
+        args = [self._do_chiral_sss]
+        logging.info('Running SQL command: %s',
+                     cursor.mogrify(command, args).decode())
+        cursor.execute(command, args)
+        command = sql.SQL('SET rdkit.tanimoto_threshold=%s')
+        args = [self._tanimoto_threshold]
+        logging.info('Running SQL command: %s',
+                     cursor.mogrify(command, args).decode())
+        cursor.execute(command, args)
+
+    def _get_tables(self):
+        """Identifies the minimum set of tables to join for the query."""
+        tables = []
+        requires_inputs = False
+        requires_rdk_inputs = False
+        requires_outputs = False
+        requires_rdk_outputs = False
+        for predicate in self._predicates:
+            if predicate.table == 'inputs':
+                if predicate.mode == ReactionComponentPredicate.MatchMode.EXACT:
+                    requires_inputs = True
+                else:
+                    requires_rdk_inputs = True
+            elif predicate.table == 'outputs':
+                if predicate.mode == ReactionComponentPredicate.MatchMode.EXACT:
+                    requires_outputs = True
+                else:
+                    requires_rdk_outputs = True
+        if requires_inputs:
+            tables.append(
+                sql.SQL("""
+                    INNER JOIN inputs USING (reaction_id) """))
+        if requires_rdk_inputs:
+            tables.append(
+                sql.SQL("""
+                    INNER JOIN rdk.inputs USING (reaction_id) """))
+        if requires_outputs:
+            tables.append(
+                sql.SQL("""
+                    INNER JOIN outputs USING (reaction_id) """))
+        if requires_rdk_outputs:
+            tables.append(
+                sql.SQL("""
+                    INNER JOIN rdk.outputs USING (reaction_id) """))
+        return tables
+
+    def run(self, cursor, limit=None):
+        """Runs the query.
+
+        Args:
+            cursor: psycopg.cursor instance.
+            limit: Integer maximum number of matches. If None (the default), no
+                limit is set.
+
+        Returns:
+            Dict mapping reaction IDs to serialized Reaction protos.
+        """
+        if not self._predicates:
+            return {}
+        self._setup(cursor)
+        components = [
+            sql.SQL("""
+            SELECT DISTINCT reaction_id, serialized
+            FROM reactions """)
+        ]
+        components.extend(self._get_tables())
+        components.append(sql.SQL("""
+            WHERE """))
```

Comment:
whitespace? Is the goal to come out later with a pretty-printed query string? Where can we see that?

Reply:
There is logging here that shows the complete (and variable-substituted) string; if you run the tests it shows up in the console.